### PR TITLE
New: Allow extends to be an array (fixes #2699)

### LINKED ISSUE #2699: allow `extends` to be an array
Currently, [extends](http://eslint.org/docs/user-guide/configuring#extending-configuration-files) only allows one file but it should allow an array:

**Sample `.eslintrc` file:**
```json
"extends" : [
	"./node_modules/coding-conventions/.eslintrc-defaults",
	"./node_modules/coding-conventions/.eslintrc-jsx",
	"./node_modules/coding-conventions/.eslintrc-es6"
]
```

### PR DESCRIPTION
**Pull request by fakewaffle**

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -95,26 +95,35 @@ function loadConfig(filePath) {
...
 
-            var parentPath = config.extends;
-            if (isFilePath(parentPath)) {
-                // If the `extends` path is relative, use the directory of the current configuration
-                // file as the reference point. Otherwise, use as-is.
-                parentPath = (!isAbsolutePath(parentPath) ?
-                    path.join(path.dirname(filePath), parentPath) :
-                    parentPath
-                );
+            if (config.extends.constructor !== Array) {
```
Comment (gcochard) [inline lib/config.js]: Should this be using `instanceof` instead? See [.constructor](https://github.com/eslint/eslint/search?p=4&q=constructor&utf8=%E2%9C%93) vs [instanceof](https://github.com/eslint/eslint/search?p=4&q=instanceof&utf8=%E2%9C%93)

Comment (gcochard) [inline lib/config.js]: Or better yet, `if (!Array.isArray(config.extends)) {...}`

**Comment by fakewaffle** [comment]
@gcochard Thank you for the comments. I have updated and amended my commit.

File: lib/config.js [hunk 2]
```diff
@@ -95,26 +95,36 @@ function loadConfig(filePath) {
...
-                config = util.mergeConfigs(loadConfig(parentPath), config);
-            } catch (e) {
-                // If the file referenced by `extends` failed to load, add the path to the
-                // configuration file that referenced it to the error message so the user is
-                // able to see where it was referenced from, then re-throw
-                e.message += "\nReferenced from: " + filePath;
-                throw e;
-            }
+            // Make the last element in an array take the highest precedence
+            config.extends = config.extends.reverse();
```
Comment (ilyavolodin) [inline lib/config.js]: reverse + forEach can be replaced with reduceRight. But I'm not sure why you do reverse anyways, that means that last element is going to get merged first, and then each consecutive element will override it, so it is, in fact lowest precedence. So in that case you should probably just use `reduce`

File: tests/lib/config.js [hunk 3]
```diff
@@ -761,6 +761,19 @@ describe("Config", function() {
             }, /Referenced from:.*?error\.json/);
         });
 
+        // Keep order with the last array element taking highest precedence
+        it("should make the last element in an array take the highest precedence", function() {
```
Comment (ilyavolodin) [inline tests/lib/config.js]: Hmmm.. I see that the unittest is passing, even though I would expect it to fail and value of `comma-dangle` to be 1... Strange, reverse should've merged in `.eslintrc2` first, and then override it with `.eslintrc1`. So my comment above might not be valid, but I still think it's much better to use `reduceRight` in that case.

File: lib/config.js (hunk 2 above)
Comment (fakewaffle) [inline lib/config.js]: Look at the way the test fixtures are structured, and what is being expected with the overrides.

File: lib/config.js [hunk 4]
```diff
@@ -95,26 +95,36 @@ function loadConfig(filePath) {
...
-            var parentPath = config.extends;
-            if (isFilePath(parentPath)) {
-                // If the `extends` path is relative, use the directory of the current configuration
-                // file as the reference point. Otherwise, use as-is.
-                parentPath = (!isAbsolutePath(parentPath) ?
-                    path.join(path.dirname(filePath), parentPath) :
-                    parentPath
-                );
+            if (!Array.isArray(config.extends)) {
+                config.extends = [config.extends];
```
Comment (nzakas) [inline lib/config.js]: We shouldn't modify the `config` in any way. Just use a variable.

File: lib/config.js (hunk 2 above)
Comment (nzakas) [inline lib/config.js]: Agree with @ilyavolodin , `reduce` would seem to be called for here.

**Comment by fakewaffle** [comment]
Long day and I feel like I am missing something really obvious?

This fails:
```js
if (config.extends) {
	var configs = config.extends;

	if (!Array.isArray(config.extends)) {
		configs = [config.extends];
	}

	// Make the last element in an array take the highest precedence
	configs.reduceRight(function (previousValue, parentPath) {

		if (isFilePath(parentPath)) {
			// If the `extends` path is relative, use the directory of the current configuration
			// file as the reference point. Otherwise, use as-is.
			parentPath = (!isAbsolutePath(parentPath) ?
				path.join(path.dirname(filePath), parentPath) :
				parentPath
			);
		}

		// Merge the configuration, ensuring children get preference over the parent
		try {
			config = util.mergeConfigs(loadConfig(parentPath), config);
		} catch (e) {
			// If the file referenced by `extends` failed to load, add the path to the
			// configuration file that referenced it to the error message so the user is
			// able to see where it was referenced from, then re-throw
			e.message += "\nReferenced from: " + filePath;
			throw e;
		}
	});

}
```

```bash


  Config
    new Config()

      ✓ should not modify baseConfig when format is specified
    findLocalConfigFiles()

      ✓ should return the path when an .eslintrc file is found

      ✓ should return an empty array when an .eslintrc file is not found

      ✓ should return the path when a package.json file is found

      ✓ should return an empty array when a package.json file is not found
    getConfig()

      ✓ should return the project config when called in current working directory

      ✓ should not retain configs from previous directories when called multiple times

      ✓ should return a default config when an invalid path is given

      ✓ should throw error when an invalid configuration file is read

      ✓ should cache config when the same directory is passed twice

      ✓ should load the config file when there are JS-style comments in the text

      ✓ should load the config file when a YAML file is used

      ✓ should return a blank config when using reset and no .eslintrc

      ✓ should return a blank config when baseConfig is set to false and no .eslintrc

      ✓ should return the default config when not using .eslintrc

      ✓ should return a modified config when baseConfig is set to an object and no .eslintrc

      ✓ should merge environment rules and ecmaFeatures into config when using .eslintrc and the built-in defaults

      ✓ should not merge environment rules into config when using .eslintrc and reset

      ✓ should merge configs when local .eslintrc overrides parent .eslintrc

      ✓ should merge configs when local .eslintrc overrides parent and grandparent .eslintrc

      ✓ should merge command line config when config file adds to local .eslintrc

      ✓ should merge command line config when config file overrides local .eslintrc

      ✓ should merge command line config when config file adds to local and parent .eslintrc

      ✓ should merge command line config when config file overrides local and parent .eslintrc

      ✓ should merge command line config when config file adds environment to local .eslintrc

      ✓ should merge command line config and rule when rule and config file overrides local .eslintrc

      ✓ should load user config globals

      ✓ should not load disabled environments

      ✓ should not error on fake environments

      ✓ should gracefully handle empty files

      1) should extend defined configuration file

      2) should extend package configuration

      3) should extend package configuration without prefix

      4) should extend scoped package configuration

      5) should extend scoped package configuration without prefix

      6) should extend package sub-configuration without prefix

      7) should extend scoped package sub-configuration without prefix

      8) should extend package configuration with sub directories

      9) should extend recursively defined configuration files

      10) should include references to where an `extends` configuration was loaded from

      11) should make the last element in an array take the highest precedence
      with plugin configuration

        ✓ should return config with plugin config

        ✓ should only return plugin config for rules with correct prefix

        ✓ should return merged configs for both plugins

        ✓ should still work if the plugin does not provide a default configuration

        ✓ should not clobber config objects when loading shared configs


  35 passing (171ms)
  11 failing

  1) Config getConfig() should extend defined configuration file:

      AssertionError: expected { Object (quotes, yoda, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

           2
           "double"
         ]
         "valid-jsdoc": 0
      -  "yoda": 2
       }
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:564:13)

  2) Config getConfig() should extend package configuration:

      AssertionError: expected { Object (quotes, foo, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "foo": "bar"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:586:13)

  3) Config getConfig() should extend package configuration without prefix:

      AssertionError: expected { Object (quotes, foo, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "foo": "bar"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:607:13)

  4) Config getConfig() should extend scoped package configuration:

      AssertionError: expected { Object (quotes, foo, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "foo": "bar"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:628:13)

  5) Config getConfig() should extend scoped package configuration without prefix:

      AssertionError: expected { Object (quotes, foo, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "foo": "bar"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:649:13)

  6) Config getConfig() should extend package sub-configuration without prefix:

      AssertionError: expected { Object (quotes, bar, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "bar": "baz"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:678:13)

  7) Config getConfig() should extend scoped package sub-configuration without prefix:

      AssertionError: expected { Object (quotes, bar, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "bar": "baz"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:707:13)

  8) Config getConfig() should extend package configuration with sub directories:

      AssertionError: expected { Object (quotes, foo, ...) } to deeply equal { Object (quotes, valid-jsdoc) }
      + expected - actual

       {
      -  "foo": "bar"
         "quotes": [
           2
           "double"
         ]
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:738:13)

  9) Config getConfig() should extend recursively defined configuration files:

      AssertionError: expected { semi: 2, yoda: 2, 'valid-jsdoc': 0 } to deeply equal { yoda: 2 }
      + expected - actual

       {
      -  "semi": 2
      -  "valid-jsdoc": 0
         "yoda": 2
       }
      
      at assertConfigsEqual (tests/lib/config.js:48:16)
      at Context.<anonymous> (tests/lib/config.js:751:13)

  10) Config getConfig() should include references to where an `extends` configuration was loaded from:
     AssertionError: expected [Function] to throw an error
      at Context.<anonymous> (tests/lib/config.js:758:20)

  11) Config getConfig() should make the last element in an array take the highest precedence:

      AssertionError: expected { Object (browser, node, ...) } to deeply equal { browser: false, node: true }
      + expected - actual

       {
         "browser": false
      -  "es6": true
         "node": true
       }
      
      at assertConfigsEqual (tests/lib/config.js:40:16)
      at Context.<anonymous> (tests/lib/config.js:774:13)
```

**Comment by ilyavolodin** [comment]
Don't merge them into `config`, merge them into `previousValue`, then return `previousValue`. After `reduceRight` is complete, merge the result into config. Also, pass second param into `reduceRight` of `{}`. Or you can pass `config` and then not merge after completion.

**Comment by fakewaffle** [comment]
Amended. It's probably worse lol. But, tests pass!

**Pull request merged**